### PR TITLE
refactor: retry failing API calls on 500 error

### PR DESCRIPTION
**Pull request by Abhikumar98**
Adding a 1 retry policy for API calls with 500 error code

### CODE REVIEW

File: src/utils/errors.ts [hunk 1]
```diff
@@ -200,5 +201,6 @@ export class HTTPError extends Error {
 
     this.name = 'HTTPError'
     this.response = response
+    this.status = response.status
```
Comment (Abhikumar98) [inline src/utils/errors.ts:204]: replace the `response.status` with `response.code`

**Review by Addminus — CHANGES_REQUESTED** [review]

**Comment by Addminus** [comment]
retry logic seems fine

**Comment by Addminus** [comment]
solution looks alright to me, @chybisov pls check if you find something

**Comment by chybisov** [comment]
Do we want to have a delay between retries? ~~This looks so wrong to me. Maybe we should fix the API in the first place and not make a second request in the hope of success~~ 😅

**Comment by Abhikumar98** [comment]
> Do we want to have a delay between retries? ~This looks so wrong to me. Maybe we should fix the API in the first place and not make a second request in the hope of success~ 😅

haha, these are for the times if the DEX responses are throwing 500.

I'll add a delay, how does 500ms sound?

**Comment by chybisov** [comment]
> > Do we want to have a delay between retries? ~This looks so wrong to me. Maybe we should fix the API in the first place and not make a second request in the hope of success~ 😅
> 
> haha, these are for the times if the DEX responses are throwing 500.
> 
> I'll add a delay, how does 500ms sound?

Yeah, but I mean backend should handle this 🙂

**Comment by Addminus** [comment]
I fear that moving this retry logic to the backend might result in timeout errors in the frontend as the logic of checking some status enpoint from dexes and bridges is expensive.

**Comment by Abhikumar98** [comment]
> I fear that moving this retry logic to the backend might result in timeout errors in the frontend as the logic of checking some status enpoint from dexes and bridges is expensive and takes time. Adding a delay is a good idea, 500ms is a good starting point.

yep, this does makes sense. Requests can result in timeout, in this case does it makes sense to have the retry logic in place just for the api calls related to DEX and bridges endpoints?

wdyt @chybisov @Addminus ?

PS: Trying to be sure of logic or any other change involved before making the fetch changes overall in the codebase.

**Comment by chybisov** [comment]
> > I fear that moving this retry logic to the backend might result in timeout errors in the frontend as the logic of checking some status enpoint from dexes and bridges is expensive and takes time. Adding a delay is a good idea, 500ms is a good starting point.
> 
> yep, this does makes sense. Requests can result in timeout, in this case does it makes sense to have the retry logic in place just for the api calls related to DEX and bridges endpoints?
> 
> wdyt @chybisov @Addminus ?
> 
> PS: Trying to be sure of logic or any other change involved before making the fetch changes overall in the codebase.

Let's create a wrapper function for all our fetch requests as per the initial plan not to overthink this. And please give it some small general name like `request` because we can decide to add some other functionality to it in the future (sorry, `lifiFetch` looks ugly to me 😅).

**Comment by Abhikumar98** [comment]
> > > I fear that moving this retry logic to the backend might result in timeout errors in the frontend as the logic of checking some status enpoint from dexes and bridges is expensive and takes time. Adding a delay is a good idea, 500ms is a good starting point.
> > 
> > yep, this does makes sense. Requests can result in timeout, in this case does it makes sense to have the retry logic in place just for the api calls related to DEX and bridges endpoints?
> > 
> > wdyt @chybisov @Addminus ?
> > 
> > PS: Trying to be sure of logic or any other change involved before making the fetch changes overall in the codebase.
> 
> Let's create a wrapper function for all our fetch requests as per the initial plan not to overthink this. And please give it some small general name like `request` because we can decide to add some other functionality to it in the future (sorry, `lifiFetch` looks ugly to me 😅). 

Cool got it. Variable naming isn't my best game so it's cool, haha.

File: src/services/ApiService.unit.spec.ts [hunk 2]
```diff
@@ -140,7 +140,7 @@ describe('ApiService', () => {
           await expect(ApiService.getRoutes(request)).rejects.toThrowError(
             new ServerError('Oops')
           )
-          expect(mockedFetch).toHaveBeenCalledTimes(1)
+          expect(mockedFetch).toHaveBeenCalledTimes(2)
```
[the remark this answers is not in the document]
Comment (Abhikumar98) [reply src/services/ApiService.unit.spec.ts:143]: it's a ServerError with a errorCode of 500. This PR deals with changes which would call this API twice being 500 error

[the remark this answers is not in the document]
Comment (Abhikumar98) [reply src/services/ApiService.unit.spec.ts:143]: There isn't a common argument interface being shared among the functions, hence it's gonna be an additional argument in a handful of them while others would be done by adding a variable in the argument interface.


We are anyways mocking the `fetch` in this scenario and afaik an actual network wouldn't be made in this case.

[the remark this answers is not in the document]
Comment (chybisov) [reply src/services/ApiService.unit.spec.ts:143]: No, we should not add any new arguments here. We need to define request settings and adjust them based on the environment - in our case tests. 

Simplified example:

1. Declare `const requestOptions =  { retries: 1 };`
2. In `beforeAll` we set `requestOptions.retries = 0`
3. Run tests...
4. In `afterAll` we set it back `requestOptions.retries = 1`

Maybe you can come up with something better, but this is the way I'd like us to go 😄

**Review by chybisov — APPROVED** [review]

**Pull request merged**